Add a new page on "Concurrency and isolates" (#9777)
flutter/website: src/perf/isolates.md

---

> title: Concurrency and Isolates

Reviewer: ```suggestion
title: Concurrency and isolates
```

description: Multithreading in Flutter using Dart isolates.
---

<?code-excerpt path-base="development/concurrency/isolates/"?>

All Dart code runs in [isolates]({{site.dart-site}}/language/concurrency), which
are similar to threads, but differ in that isolates have their own isolated
memory. They do not share state in any way, and can only communicate by
messaging. By default, Flutter apps do all of their work on a single isolate –
the main isolate. In most cases, this model allows for simpler programming and
is fast enough that the application's UI doesn’t become unresponsive.

Sometimes, though, applications need to perform exceptionally large computations

> that can cause “UI jank” (jerky motion). If your app is experiencing jank for

Reviewer: ```suggestion
that can cause "UI jank" (jerky motion). If your app is experiencing jank for
```

> this reason, you can move these computations to a helper isolate; allowing the
> underlying runtime environment to run the computation concurrently with the main

Reviewer: ```suggestion
this reason, you can move these computations to a helper isolate. This allows the
underlying runtime environment to run the computation concurrently with the main
```

UI isolate's work and takes advantage of multi-core devices.

Each isolate has its own memory and its own event loop. The event loop processes

> events in the order that they’re added to an event queue. On the main isolate,

Reviewer: ```suggestion
events in the order that they're added to an event queue. On the main isolate,
```

these events can be anything from handling a user tapping in the UI, to
executing a function, to painting a frame on the screen. The following figure
shows an example event queue with 4 events waiting to be processed.

![The main isolate diagram]({{site.url}}/assets/images/docs/development/concurrency/basics-main-isolate.png){:width="50%"}

> For smooth rendering, Flutter adds a “paint frame” event to the event queue 60

Reviewer: ```suggestion
For smooth rendering, Flutter adds a "paint frame" event to the event queue 60
```

times per second (for a 60Hz device). If these events aren’t processed on time,
the application experiences UI jank, or worse, become unresponsive altogether.

![Event jank diagram]({{site.url}}/assets/images/docs/development/concurrency/event-jank.png){:width="50%"}

Whenever a process can’t be completed in a frame gap, or the time between two
frames, it's a good idea to offload the work to another isolate to ensure that
the main isolate can produce 60 frames per second. When you spawn an isolate in
Dart, it can process the work concurrently with the main isolate, without
blocking it.

You can read more about how isolates and the event loop work in Dart on
the [concurrency page][] of the Dart
documentation.

[concurrency page]: {{site.dart-site}}/language/concurrency

<iframe width="560" height="315" src="{{site.youtube-site}}/embed/vl_AaCgudcY" 
 frameborder="0" allow="accelerometer; autoplay;encrypted-media; gyroscope; picture-in-picture" 
 allowfullscreen></iframe>

## Message passing between isolates

> Dart’s isolates are an implementation of the [Actor model][]. They can only

Reviewer: ```suggestion
Dart's isolates are an implementation of the [Actor model][]. They can only
```

> communicate with each other by message passing. When messages are “passed”

Reviewer: ```suggestion
communicate with each other by message passing. When messages are "passed"
```

between each other, they are generally copied from the sending isolate to the
receiving isolate. This means that any value passed to an isolate, even if

> mutated on that isolate, doesn’t change the value on the original isolate.

Reviewer: ```suggestion
mutated on that isolate, doesn't change the value on the original isolate.
```

> The only objects that aren’t copied when passed to an isolate are immutable

Reviewer: OK, this is a global change - I won't keep calling it out from here on. Change the double and single quotes to the straight up and down kind.  :)

```suggestion
The only objects that aren't copied when passed to an isolate are immutable
```

objects that can’t be changed anyway, such a String or an unmodifiable byte.
When you pass an immutable object between isolates, it is transferred, rather
than copied, in order to save CPU memory. Because immutable objects can’t be
updated, this effectively retains the actor model behavior.

An exception to this rule is when an isolate exits when it sends a message using

> the `Isolate.exit` method. Because the sending isolate will not exist after

Reviewer: ```suggestion
the `Isolate.exit` method. Because the sending isolate won't exist after
```

sending the message, it can pass the reference to the message while ensuring
that only one isolate can access the message.

The two lowest-level primitives that send messages are `SendPort.send`, which

> will make a copy of a mutable message as it sends, and `Isolate.exit`, which
> will send the reference to the message. Both `Isolate.run` and `compute`

Reviewer: Avoid future tense.

```suggestion
makes a copy of a mutable message as it sends, and `Isolate.exit`, which
sends the reference to the message. Both `Isolate.run` and `compute`
```

callback’s value back to the main isolate, and exits the spawned isolate.

For example, consider this code that loads a large JSON blob from a file, and
converts that JSON into custom Dart objects. If the json decoding process wasn’t
off loaded to a new isolate, this method would cause the UI to become
unresponsive for several seconds.

<?code-excerpt "lib/main.dart (isolate-run)"?>
```dart
// Produces a list of 211,640 photo objects.
// (The JSON file is ~20MB.)
Future<List<Photo>> getPhotos() async {
  String jsonString = await rootBundle.loadString('assets/photos.json');

  final dynamic photos = await Isolate.run(() {
    final photoData = jsonDecode(jsonString);
    return photoData.map((dynamic element) {
      final data = element as Map<String, dynamic>;
      return Photo.fromJson(data);
    }).toList();
  });

  return photos as List<Photo>;
}
```

For a complete walkthrough of using Isolates to parse JSON in the background, see [this cookbook recipe][].

[this cookbook recipe]: {{site.url}}/cookbook/networking/background-parsing

### Isolates on the web

Dart doesn’t support isolates on web platforms. If you’re targeting the web with

> your Flutter app, you should use the compute method. The [`compute()`][] method

Reviewer: ```suggestion
your Flutter app, you should use the `compute` method. The [`compute()`][] method
```

runs the computation on the main thread on the web, but spawns a new thread on
mobile devices. On mobile and desktop platforms `await compute(fun, message)` is
equivalent to `await Isolate.run(() => fun(message))`.

## Stateful, longer-lived isolates

> 

Reviewer: ```suggestion
Please add a sentence introducing this list. Such as "The following list describes longer-lived isolates:"
```

- [`compute()`][] abstracts a handful of lower-level, isolate-related APIs to  simplify isolate management:
- [`Isolate.spawn()`][] and [`Isolate.exit()`][]
- [`ReceivePort`][] and [`SendPort`][]
- [`send()`][] method

When you use the compute method, the new isolate immediately shuts down after it
returns a single message to the main isolate. Sometimes, you’ll need isolates
that are long lived, and can pass multiple messages to each other over time. In
Dart, you can accomplish this with the Isolate API and Ports. These long-lived
isolates are colloquially known as _background workers_.

Long-lived isolates are useful when you have a specific process that either
needs to be run repeatedly throughout the lifetime of your application, or if
you have a process that runs over a period of time and needs to yield multiple
return values to the main isolate.

### ReceivePorts and SendPorts

Set up long-lived communication between isolates with two classes (in addition
to Isolate): [`ReceivePort`][] and [`SendPort`][]. These ports are the only way isolates can
communicate with each other.

> Ports behave similarly to Streams, in which the StreamController or Sink is

Reviewer: Generally, if we uppercase the class and are talking about a specific class, we code font it. If we don't uppercase it, and are using the word generically, we don't code font it.

```suggestion
`Ports` behave similarly to `Streams`, in which the `StreamController` or `Sink` is
```

created in one isolate, and the listener is set up in the other isolate. In this

> analogy, the StreamConroller is called a SendPort, and you can “add” messages
> with the send() method. ReceivePorts are the listeners, and when these listeners

Reviewer: ```suggestion
analogy, the `StreamConroller` is called a `SendPort`, and you can "add" messages
with the `send()` method. `ReceivePort`s are the listeners, and when these listeners
```

argument.

For an in-depth explanation on setting up two-way communication between the main
isolate and a worker isolate, follow the examples in the [Dart documentation][].

[Dart documentation]: {{site.dart-site}}/language/concurrency

## Common use cases for isolates

There is only one hard rule for when you should use isolates, and that’s when
large computations are causing your Flutter application to experience UI jank.
This jank happens when there is any computation that takes longer than the
Flutter’s frame gap.

![Event jank diagram]({{site.url}}/assets/images/docs/development/concurrency/event-jank.png){:width="50%"}

Any process _could_ take longer to complete, depending on the implementation and the input data, making it impossible to create an exhaustive list of when you need to consider using isolates.

That said, isolates are commonly used for the following:

- Reading data from a local database
- Sending push notifications
- Parsing and decoding large data files
- Processing or compressing photos, audio files, and video files
- Converting audio and video files
- When you need asynchronous support while using FFI

> - Applying filtering to complex lists or filesystems

Reviewer: I can't help but think that this very useful list should appear MUCH higher in this page.

### Using platform plugins in isolates

As of Flutter 3.7, you can use platform plugins in background isolates. This
opens many possibilities to offload heavy, platform-dependent computations to an
isolate that won’t block your UI. For example, imagine you’re encrypting data
using a native host API (such as an Android API on Android, an iOS API on iOS,
and so on). Previously, [marshaling data][] to the host platform could waste UI
thread time, and can now be done in a background isolate.

> Platform channel isolates use the [BackgroundIsolateBinaryMessenger][] API. The

Reviewer: ```suggestion
Platform channel isolates use the [`BackgroundIsolateBinaryMessenger`][] API. The
```

their own global fields, and can only communicate with message passing, ensuring
that mutable objects in an isolate are only ever accessible in a single isolate.
Therefore, isolates are limited by their access to their own memory. For
example, if you have an application with a global mutable variable
called `configuration`, it will be copied as a new global field in a spawned
isolate. If you mutate that variable in the spawned isolate, it will remain
untouched in the main isolate. This is true even if you pass the `configuration`
object as a message to the new isolate. This is how isolates are meant to
function, and it's important to keep in mind when you’re considering using
isolates.

### Web platforms

Dart web platforms, including Flutter web, don’t support isolates.
Using `compute` in a web application calls the callback on the main isolate. For
more information, check out the [concurrency documentation][] on dart.dev.

[concurrency documentation]: {{site.dart-site}}/language/concurrency

### No `rootBundle` access or `dart:ui` methods

All UI tasks and Flutter itself are coupled to the main isolate. Therefore, you

> can’t access assets using rootBundle in spawned isolates, nor can you perform

Reviewer: ```suggestion
can't access assets using `rootBundle` in spawned isolates, nor can you perform
```

any widget or UI work in spawned isolates.

### Limited plugin messages from host platform to Flutter

> With background isolate platform channels,you can use platform channels in

Reviewer: ```suggestion
With background isolate platform channels, you can use platform channels in
```

isolates to send messages to the host platform (for example Android or iOS), and
receive responses to those messages. However, you can’t receive unsolicited
messages from the host platform.

As an example, you can’t set-up a long-lived Firestore listener in a background
isolate, because Firestore uses platform channels to push updates to Flutter,
which are unsolicited. You can, however, query Firestore for a response in the
background.

## More information

For more information on isolates, check out the following resources:

- If you’re using many isolates, consider the [IsolateNameServer][] class in Flutter,
or the pub package that clones the functionality for Dart applications not using
Flutter. 
- Dart’s Isolates are an implementation of the [Actor model][]. 

> - Read more about the BackgroundIsolateBinaryMessenger API [announcement][].

Reviewer: ```suggestion
- Read more about the `BackgroundIsolateBinaryMessenger` API [announcement][].
```

[announcement]: https://medium.com/flutter/introducing-background-isolate-channels-7a299609cad8
[technical design proposal]: https://docs.google.com/document/d/1yAFw-6kBefuurXWTur9jdEUAckWiWJVukP1Iay8ehyU/edit#heading=h.722pnbmlqbkx

[Actor model]: https://en.wikipedia.org/wiki/Actor_model
[marshaling data]: https://en.wikipedia.org/wiki/Marshalling_(computer_science)
[Dart `Isolate`]: {{site.dart.api}}/stable/dart-isolate/Isolate-class.html
[`compute()`]: {{site.api}}/flutter/foundation/compute.html
[`Isolate.spawn()`]: {{site.dart.api}}/stable/dart-isolate/Isolate/spawn.html
[`Isolate.exit()`]: {{site.dart.api}}/stable/dart-isolate/Isolate/exit.html
[`ReceivePort`]: {{site.dart.api}}/stable/dart-isolate/ReceivePort-class.html
[`SendPort`]: {{site.dart.api}}/stable/dart-isolate/SendPort-class.html
[`send()`]: {{site.dart.api}}/stable/dart-isolate/SendPort/send.html

> [BackgroundIsolateBinaryMessenger]: {{site.api}}/flutter/services/BackgroundIsolateBinaryMessenger-class.html

Reviewer: ```suggestion
[`BackgroundIsolateBinaryMessenger`]: {{site.api}}/flutter/services/BackgroundIsolateBinaryMessenger-class.html
```

All Dart code runs in [isolates]({{site.dart-site}}/language/concurrency),
which are similar to threads,
but differ in that isolates have their own isolated memory.
They do not share state in any way,
and can only communicate by messaging.
By default,
Flutter apps do all of their work on a single isolate –
the main isolate.
In most cases, this model allows for simpler programming and
is fast enough that the application's UI doesn't become unresponsive.

Sometimes though, 
applications need to perform exceptionally large computations
that can cause "UI jank" (jerky motion).
If your app is experiencing jank for this reason,
you can move these computations to a helper isolate.
This allows the underlying runtime environment 
to run the computation concurrently
with the main UI isolate's work
and takes advantage of multi-core devices.

Each isolate has its own memory
and its own event loop.
The event loop processes
events in the order that they're added to an event queue.
On the main isolate,
these events can be anything from handling a user tapping in the UI,
to executing a function,
to painting a frame on the screen.
The following figure shows an example event queue

> with 3 events waiting to be processed.

Reviewer 2: nit: These lines are *aggressively* newline'd.  You might find it helpful to use plugins for vscode that will automatically do the wordwrap for you.  I use this one: https://marketplace.visualstudio.com/items?itemName=stkb.rewrap

Author: Yeah.. this is a practice called [Semantic Line Breaks](https://sembr.org/) that I was introduced to via previous comments on this very PR. I'm hoping someone with more knowledge about "Sembr" can tell me if I did it _too aggressively_.

![The main isolate diagram]({{site.url}}/assets/images/docs/development/concurrency/basics-main-isolate.png){:width="50%"}

For smooth rendering,
Flutter adds a "paint frame" event to the event queue
60 times per second(for a 60Hz device).
If these events aren't processed on time,
the application experiences UI jank,
or worse,
become unresponsive altogether.

![Event jank diagram]({{site.url}}/assets/images/docs/development/concurrency/event-jank.png){:width="50%"}

Whenever a process can't be completed in a frame gap,

> or the time between two frames,

Reviewer 2: I would remove the "or" since it isn't clear you are describing the frame gap like this.

 frameborder="0" allow="accelerometer; autoplay;encrypted-media; gyroscope; picture-in-picture" 
 allowfullscreen></iframe>

## Common use cases for isolates

There is only one hard rule for when you should use isolates,
and that's when large computations are causing your Flutter application
to experience UI jank.
This jank happens when there is any computation that takes longer than
Flutter's frame gap.

![Event jank diagram]({{site.url}}/assets/images/docs/development/concurrency/event-jank.png){:width="50%"}

Any process _could_ take longer to complete,
depending on the implementation
and the input data,
making it impossible to create an exhaustive list of
when you need to consider using isolates.

That said, isolates are commonly used for the following:

- Reading data from a local database
- Sending push notifications
- Parsing and decoding large data files
- Processing or compressing photos, audio files, and video files
- Converting audio and video files
- When you need asynchronous support while using FFI
- Applying filtering to complex lists or filesystems

## Message passing between isolates

Dart's isolates are an implementation of the [Actor model][].

> They can only communicate with each other by message passing.

Reviewer 2: Maybe it's worth mentioning "port"s here?

I think in general in this doc whenever it's using the "passing messages" nomenclature it might be more clear to use Dart nomenclature like "ports".

When messages are "passed" between each other,
they are generally copied from the sending isolate to the
receiving isolate.
This means that any value passed to an isolate,
even if mutated on that isolate,
doesn't change the value on the original isolate.

The only objects that aren't copied when passed to an isolate
are immutable objects that can't be changed anyway,

> such a String or an unmodifiable byte.

Reviewer 2: Can you link to the Dart documentation on these rules?

When you pass an immutable object between isolates,

> it is transferred, rather than copied,

Reviewer 2: "Transferred" is ambiguous, I'd say that a reference to the data is sent.

for better performance.
Because immutable objects can't be updated,
this effectively retains the actor model behavior.

An exception to this rule is
when an isolate exits when it sends a message using the `Isolate.exit` method.
Because the sending isolate won't exist after sending the message,

> it can pass the reference to the message while

Reviewer 2: I'd say it passes ownership of the message from one isolate to another.

ensuring that only one isolate can access the message.

The two lowest-level primitives that send messages are `SendPort.send`,
which makes a copy of a mutable message as it sends,
and `Isolate.exit`,
which sends the reference to the message.
Both `Isolate.run` and `compute`
use `Isolate.exit` under the hood.

## Short-lived isolates

> The easiest way to move a process to an isolate in Flutter is with

Reviewer 2: It may be worth mentioning the overhead associated copying messages and spawning isolates so people get some intuition about the tradeoffs.

and Ports.
These long-lived isolates are colloquially known as _background workers_.

Long-lived isolates are useful when you have a specific process that either
needs to be run repeatedly throughout the lifetime of your application,
or if you have a process that runs over a period of time
and needs to yield multiple return values to the main isolate.

### ReceivePorts and SendPorts

Set up long-lived communication between isolates with two classes
(in addition to Isolate):
[`ReceivePort`][] and [`SendPort`][].
These ports are the only way isolates can communicate with each other.

`Ports` behave similarly to `Streams`,
in which the `StreamController`
or `Sink` is created in one isolate,
and the listener is set up in the other isolate.
In this analogy,
the `StreamConroller` is called a `SendPort`,
and you can "add" messages with the `send()` method.
`ReceivePort`s are the listeners,
and when these listeners receive a new message,
they call a provided callback with the message as an argument.

For an in-depth explanation on setting up two-way
communication between the main isolate
and a worker isolate,
follow the examples in the [Dart documentation][].

> [Dart documentation]: {{site.dart-site}}/language/concurrency

Reviewer 2: FWIW I have an easier way https://medium.com/@gaaclarke/isolate-agents-easy-isolates-for-flutter-6d75bf69a2e7.

## Using platform plugins in isolates

As of Flutter 3.7, you can use platform plugins in background isolates.
This opens many possibilities to offload heavy,
platform-dependent computations to an isolate that won't block your UI.
For example, imagine you're encrypting data
using a native host API
(such as an Android API on Android, an iOS API on iOS, and so on).
Previously, [marshaling data][] to the host platform could waste UI thread time,
and can now be done in a background isolate.

Platform channel isolates use the [`BackgroundIsolateBinaryMessenger`][] API.
The following snippet shows an example of using 

> the shared_preferences package in a background isolate.

Reviewer 2: ```suggestion
the `shared_preferences` package in a background isolate.
```

<?code-excerpt "lib/isolate_binary_messenger.dart"?>
```dart
import 'dart:isolate';

import 'package:flutter/services.dart';
import 'package:shared_preferences/shared_preferences.dart';

void main() {
  // Identify the root isolate to pass to the background isolate.
  RootIsolateToken rootIsolateToken = RootIsolateToken.instance!;
  Isolate.spawn(_isolateMain, rootIsolateToken);
}

Future<void> _isolateMain(RootIsolateToken rootIsolateToken) async {
  // Register the background isolate with the root isolate.
  BackgroundIsolateBinaryMessenger.ensureInitialized(rootIsolateToken);

  // You can now use the shared_preferences plugin.
  SharedPreferences sharedPreferences = await SharedPreferences.getInstance();

  print(sharedPreferences.getBool('isDebug'));
}
```

## Limitations of Isolates

If you're coming to Dart from a language with multithreading,
it's reasonable to expect isolates to behave like threads,
but that isn't the case.
Isolates have their own global fields,

> and can only communicate with message passing,

Reviewer 2: The nomenclature of this document is "message passing", I wonder if it would be better to use the nomenclature of Dart which is "ports".

All Dart code runs in [isolates]({{site.dart-site}}/language/concurrency),
which are similar to threads,
but differ in that isolates have their own isolated memory.
They do not share state in any way,
and can only communicate by messaging.
By default,
Flutter apps do all of their work on a single isolate –
the main isolate.
In most cases, this model allows for simpler programming and
is fast enough that the application's UI doesn't become unresponsive.

Sometimes though, 
applications need to perform exceptionally large computations
that can cause "UI jank" (jerky motion).
If your app is experiencing jank for this reason,
you can move these computations to a helper isolate.
This allows the underlying runtime environment 
to run the computation concurrently
with the main UI isolate's work
and takes advantage of multi-core devices.

Each isolate has its own memory
and its own event loop.
The event loop processes
events in the order that they're added to an event queue.
On the main isolate,
these events can be anything from handling a user tapping in the UI,
to executing a function,
to painting a frame on the screen.
The following figure shows an example event queue

> with 4 events waiting to be processed.

Reviewer 3: I only saw 3 green events in the image. Unless you were referring to the blue rectangles?

When the computation completes,
it returns the callback's value back to the main isolate,
and exits the spawned isolate.

For example,
consider this code that loads a large JSON blob from a file,
and converts that JSON into custom Dart objects.
If the json decoding process wasn't off loaded to a new isolate,
this method would cause the UI to
become unresponsive for several seconds.

<?code-excerpt "lib/main.dart (isolate-run)"?>
```dart
// Produces a list of 211,640 photo objects.
// (The JSON file is ~20MB.)
Future<List<Photo>> getPhotos() async {
  String jsonString = await rootBundle.loadString('assets/photos.json');

  final dynamic photos = await Isolate.run(() {
    final photoData = jsonDecode(jsonString);
    return photoData.map((dynamic element) {
      final data = element as Map<String, dynamic>;
      return Photo.fromJson(data);
    }).toList();
  });

  return photos as List<Photo>;
}
```
For a complete walkthrough of using Isolates to 
parse JSON in the background, see [this cookbook recipe][].

[this cookbook recipe]: {{site.url}}/cookbook/networking/background-parsing

### Isolates on the web

Dart doesn't support isolates on web platforms.

> If you're targeting the web with your Flutter app,

Reviewer 3: Having to recommend `compute()` on the web, where it runs things on the main thread anyway, seems kind of defeating the point of offloading work onto a different thread/isolate.

Instead, you can refer to the `WebWorker` API, which is actually quite similar to the Isolate API (start/stop a WebWorker, passing messages & transferable by-reference data types).

We have first-party support in [package:web](https://github.com/dart-lang/web/blob/e99d53716709d1cbce5533a7f6d036a8a5e19f68/lib/src/dom/html.dart#L4782) You could also link to https://developer.mozilla.org/en-US/docs/Web/API/Web_Workers_API

Perhaps a code sample where you write the photo sample with a WebWorker could be nice to have?

Author: Overwhelmingly, the most requested addition to our docs about isolates is being clear about the limitations. We have resources sprinkled across the web that use `compute` in example code, and it's important that we're explicit about what that means. 

I think documenting WebWorkers is beyond the scope of this update.

Reviewer 4: I agree we don't need to document web workers here, but we do need to be careful about how we write about `compute` and its behavior on web.

Some people incorrectly see it as "using `compute` instead of `Isolate.run` allows the same behavior to work on web".

I wonder if we should just always talk about and default to `Isolate.run`, but then in a small note or similar put forward `compute` as a way that allows the same code to **compile** for web without changes, but highlight that it does **not work** the same. That way it's a more explicit choice they have to make after they know the differences.
